subsys/net/ip/tcp.c

```diff
@@ -3088,10 +3097,19 @@ int net_tcp_connect(struct net_context *context,
 		conn->dst.sin6.sin6_port = remote_port;
 		conn->src.sin6.sin6_port = local_port;
 
-		ip6 = net_if_ipv6_select_src_addr(
+		if (net_ipv6_is_addr_unspecified(
+			&net_sin6(local_addr)->sin6_addr))
+		{
+			ip6 = net_if_ipv6_select_src_addr(
 					net_context_get_iface(context),
 					&net_sin6(remote_addr)->sin6_addr);
-		conn->src.sin6.sin6_addr = *ip6;
+			conn->src.sin6.sin6_addr = *ip6;
+		}
+		else
+		{
+			net_ipaddr_copy(&conn->src.sin6.sin6_addr,
+				&net_sin6(local_addr)->sin6_addr);
```

Comment:
nit, but it's good to align arguments to the opening brackets (like in the `net_ipaddr_copy()` just below).